File: java-to-go-projects/06-json-config-merger/Main.java
package merger;

import com.fasterxml.jackson.databind.JsonNode;

/**
 * CLI entry point for the JSON config merger.
 *
 * Usage: java Main base.json override1.json [override2.json ...]
 *
 * Merges multiple config files left-to-right. Later files override earlier ones.
 *
 * Go structure:
 *   main.go              — CLI parsing, orchestration
 *   merger/merge.go      — Merge(base, override) using map[string]interface{}
 *   merger/loader.go     — LoadFile, PrettyPrint
 *   merger/merge_test.go — Table-driven tests for merge edge cases
 *
 * Rust structure:
 *   main.rs              — CLI parsing
 *   merger.rs            — merge(Value, Value) -> Value using serde_json::Value
 *   loader.rs            — load_file, pretty_print
 */
public class Main {

    public static void main(String[] args) throws Exception {
        if (args.length < 2) {
            System.err.println("Usage: config-merger <base.json> <override.json> [more...]");
            System.exit(1);
        }

        JsonNode result = Loader.loadFile(args[0]);
        System.out.println("Base: " + args[0]);

        for (int i = 1; i < args.length; i++) {
            System.out.println("Merging: " + args[i]);
            JsonNode override = Loader.loadFile(args[i]);
            result = Merger.merge(result, override);
        }

        System.out.println("\nResult:");
        System.out.println(Loader.prettyPrint(result));
    }
}

File: java-to-go-projects/06-json-config-merger/go/main.go
package jsonconfigmerger
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

// loadFile reads and parses a JSON file into a generic interface value.
func loadFile(filename string) (interface{}, error) {
	panic("todo")
}

// merge deep-merges two JSON values.
// Objects: merge recursively (override keys win).
// Arrays: replaced entirely.
// Scalars: override replaces base.
// Null in override removes the key.
func merge(base, override interface{}) interface{} {
	panic("todo")
}

// prettyPrint returns an indented JSON string.
func prettyPrint(value interface{}) (string, error) {
	_ = json.MarshalIndent
	panic("todo")
}

func main() {
	if len(os.Args) < 3 {
		fmt.Fprintf(os.Stderr, "Usage: %s <base.json> <override1.json> [override2.json ...]\n", os.Args[0])
		os.Exit(1)
	}

	panic("todo: load files, merge left-to-right, pretty-print result")
}
